tests/robustness/report/wal.go

```diff
@@ -0,0 +1,237 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package report
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/go-cmp/cmp"
+	"go.uber.org/zap"
+
+	pb "go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.etcd.io/etcd/pkg/v3/pbutil"
+	"go.etcd.io/etcd/server/v3/storage/wal"
+	"go.etcd.io/etcd/server/v3/storage/wal/walpb"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+	"go.etcd.io/etcd/tests/v3/robustness/model"
+	"go.etcd.io/raft/v3/raftpb"
+)
+
+func LoadClusterPersistedRequests(lg *zap.Logger, path string) ([]model.EtcdRequest, error) {
+	files, err := os.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+	dataDirs := []string{}
+	for _, file := range files {
+		if file.IsDir() && strings.HasPrefix(file.Name(), "server-") {
+			dataDirs = append(dataDirs, filepath.Join(path, file.Name()))
+		}
+	}
+	return PersistedRequestsDirs(lg, dataDirs)
+}
+
+func PersistedRequestsCluster(lg *zap.Logger, cluster *e2e.EtcdProcessCluster) ([]model.EtcdRequest, error) {
+	dataDirs := []string{}
+	for _, proc := range cluster.Procs {
+		dataDirs = append(dataDirs, proc.Config().DataDirPath)
+	}
+	return PersistedRequestsDirs(lg, dataDirs)
+}
+
+func PersistedRequestsDirs(lg *zap.Logger, dataDirs []string) ([]model.EtcdRequest, error) {
+	persistedRequests := []model.EtcdRequest{}
+	allowedFailures := len(dataDirs) / 2
+	for _, dir := range dataDirs {
+		memberRequests, err := requestsPersistedInWAL(lg, dir)
+		if err != nil {
+			if allowedFailures < 1 {
+				return nil, err
+			}
+			allowedFailures--
+			continue
+		}
+		minLength := min(len(persistedRequests), len(memberRequests))
+		if diff := cmp.Diff(memberRequests[:minLength], persistedRequests[:minLength]); diff != "" {
+			return nil, fmt.Errorf("unexpected differences between wal entries, diff:\n%s", diff)
+		}
+		if len(memberRequests) > len(persistedRequests) {
+			persistedRequests = memberRequests
+		}
+	}
+	return persistedRequests, nil
+}
+
+func requestsPersistedInWAL(lg *zap.Logger, dataDir string) ([]model.EtcdRequest, error) {
+	_, ents, err := ReadWAL(lg, dataDir)
+	if err != nil {
+		return nil, err
+	}
+	requests := make([]model.EtcdRequest, 0, len(ents))
+	for _, ent := range ents {
+		if ent.Type != raftpb.EntryNormal || len(ent.Data) == 0 {
+			continue
+		}
+		request, err := parseEntryNormal(ent)
+		if err != nil {
+			return nil, err
+		}
+		if request != nil {
+			requests = append(requests, *request)
+		}
+	}
+	return requests, nil
+}
+
+func ReadWAL(lg *zap.Logger, dataDir string) (state raftpb.HardState, ents []raftpb.Entry, err error) {
+	walDir := filepath.Join(dataDir, "member", "wal")
+	repaired := false
+	for {
+		w, err := wal.OpenForRead(lg, walDir, walpb.Snapshot{Index: 0})
+		if err != nil {
+			lg.Fatal("failed to open WAL", zap.Error(err))
```

Comment:
Can we bubble this up and call `t.Fatal` instead to make it more readable?

Reply:
Done